config/config-api/src/com/thoughtworks/go/config/BasicCruiseConfig.java

```diff
@@ -287,14 +224,15 @@ private PipelineGroups mergePipelineConfigs() {
 
             // first add pipeline configs from main part
             List<PipelineConfigs> allPipelineConfigs = new ArrayList<>();
-            for(PipelineConfigs partPipesConf : this.main.getGroups())
-            {
+            for (PipelineConfigs partPipesConf : this.main.getGroups()) {
                 allPipelineConfigs.add(partPipesConf);
             }
             // then add from each part
             for (PartialConfig part : this.parts) {
-                for(PipelineConfigs partPipesConf : part.getGroups())
-                {
+                for (PipelineConfigs partPipesConf : part.getGroups()) {
+                    for (PipelineConfig pipelineConfig : partPipesConf) {
+                        main.getAllPipelineConfigs().add(pipelineConfig);
```

Comment:
Because of this line, we have to be sure that this method is called only once, and we have to make sure that it is called on a config which is bare/new (doesn't have pipeline configs from partials). Can this be done elsewhere, or be more explicit? One way is to return a merge result object around here, with both groups and these configs, which can then be explicitly added to the allPipelineConfigs list. Just a thought.